### ISSUE TITLE
Files uploaded through PUT do not make it into handler intact

### ISSUE THREAD

**Issue by mpetrovic** [RESTful-Drupal/restful#483]
I'm trying to replace a file on the server with the file given by the user, and using the PUT method to do so. 

I'm setting the Content-Type header to the mimetype of the file, and the file itself as the data. By the time it gets to my Restful plugin, the data is gone. There's a few bytes of headers in the request array, and then nothing. 

Looking at restful_parse_request, it looks like its treating everything that isn't a GET or POST like it's assuming all content will either be in JSON or query string format. 

**Proposal**
Check the Content-Type header for the request when the method is not GET or POST. If it's application/x-www-form-urlencoded, then run it through parse_str. If it's application/json, run it through drupal_json_decode. If it's mulltipart/form-data, parse it manually. if it's anything else, pass the request data untouched to the handler for it to handle.

**Comment by rbran100** [comment]
@mateu-aguilo-bosch does @mpetrovic solution sound good to you?

**Comment by amitaibu** [comment]
Hi OpenScholar folks! :)
I'm not sure I follow. What I would expect to send (and I'm not sure it's going to work, as I've never tested it when I wrote the File resource), is something like this:

**Comment by amitaibu** [comment]
![postman](https://cloud.githubusercontent.com/assets/125707/7235657/11d4cebc-e798-11e4-9217-4b383c0494f8.jpg)

**Comment by mpetrovic** [comment]
I've been trying with lots of things. The closest I've gotten is with raw. It cannot be form-data, since php://input just doesn't work. I haven't tried x-www-form-urlencoded in my own stuff, so I don't know how that handles files.

**Comment by amitaibu** [comment]
@mpetrovic wouldn't another solution be:

1) Create a new file (as would happen if you used a form)
2) Replace the references in the node(s) referencing the new file
3) Delete the old file

**Comment by mpetrovic** [comment]
That is way more work then swapping the file out on disk. More error-prone too. All I'm trying to do is replace the file on disk with a new one.

**Comment by amitaibu** [comment]
Are you using the default file upload (`/api/file-upload`). If so then it currently support only [create](https://github.com/RESTful-Drupal/restful/blob/7.x-1.x/plugins/restful/RestfulFilesUpload.php#L16)

> That is way more work then swapping the file out on disk. More error-prone too

You are probably right, but I believe it's the same it's handled when uploading a file via form.

**Comment by mpetrovic** [comment]
No, I'm using our own handler. But it's irrelevant, since php://input is being read and parsed by the restful module before it invokes the handler at all. 

When using the core Replace File form found on the File Edit form, it invokes file_save_upload() with an argument of FILE_EXISTS_REPLACE. It finds the existing file with the same uri, puts the new file there, and clears caches on the old file object. No new File entity is created.

**Comment by amitaibu** [comment]
> No, I'm using our own handler.

Can you share the link so I get more context.

> it invokes file_save_upload() with an argument of FILE_EXISTS_REPLACE

Indeed, you are right :)

**Comment by mpetrovic** [comment]
https://github.com/openscholar/openscholar/tree/media_browser_rewrite/openscholar/modules/os/modules/os_restful/plugins/restful/files/1.0

**Comment by amitaibu** [comment]
I see you are using `file_save_upload()` directly. Why not use a similar code like we have [here](https://github.com/RESTful-Drupal/restful/blob/7.x-1.x/plugins/restful/RestfulFilesUpload.php#L58-L82)?

btw,  `$validators = array();  // TODO: How do we populate this?` -- this is taken from the [plugin](https://github.com/RESTful-Drupal/restful/blob/7.x-1.x/plugins/restful/RestfulFilesUpload.php#L24-L28)

**Comment by mpetrovic** [comment]
$_FILES is empty during PUT operations. 

Re: Validators.
In normal Drupal, validators take different arguments based on the Field in question. The REST api has no knowledge of what field we're trying to add this file to, which makes this approach not feasible.

**Comment by amitaibu** [comment]
> $_FILES is empty during PUT operations.

I didn't know of this limitation. Anyway, maybe use `POST`, but set the [X-HTTP-Method-Override](https://github.com/RESTful-Drupal/restful/blob/7c9bc9f76d143f6caf6d2fd293dc865b434dd799/restful.module#L420) header.

Right, the default file upload by RESTful isn't coupled with fields. So you could always move the logic you wanted to the constructor, and get the validators by `field_info_field()`.

**Comment by mpetrovic** [comment]
Why not just fix restful_parse_request() so a user can PUT a whole file and expect something in the handler?

**Comment by amitaibu** [comment]
PRs are welcome :)

**Comment by amitaibu** [comment]
(or at least some code to better understand what's your proposed solution)

**Comment by mpetrovic** [comment]
``` php
function restful_parse_request() {
  $request = &drupal_static(__FUNCTION__, NULL);
  if ($request) {
    return $request;
  }

  $method = strtoupper($_SERVER['REQUEST_METHOD']);

  if ($method == \RestfulInterface::GET) {
    $request = $_GET;
  }
  elseif ($method == \RestfulInterface::POST) {
    $request = $_POST;
  }

   // NEW CODE BEGINS HERE
  if (!$request) {
    $contentType = $_SERVER['CONTENT_TYPE'];
    switch ($contentType) {
      case 'application/x-www-form-encoded':
        parse_str('php://input', $request)
        break;
      case 'multipart/form-data':
        throw new exception('multipart/form-data and php://input do not work together')
        break;
      default:
        // either the data 
        $request['file'] = something; // either the result of file_get_contents() or fopen. Whichever makes more sense terms of memory
        break;
    }
  }

// NEW CODE ENDS HERE

  // This flag is used to identify if the request is done "via Drupal" or "via
  // CURL";
  $request['__application'] = array(
    'rest_call' => TRUE,
    'csrf_token' => \RestfulManager::getRequestHttpHeader('X-CSRF-Token'),
  );

  // Allow implementing modules to alter the request.
  drupal_alter('restful_parse_request', $request);

  return $request;
}
```

**Comment by amitaibu** [comment]
@mpetrovic I think your RESTful isn't updated - as the `php://input` is  already [used](https://github.com/RESTful-Drupal/restful/blob/7.x-1.x/restful.module#L559-L569).

Can you please check with the latest dev.

**Comment by mpetrovic** [comment]
Except the problem is [this line](https://github.com/RESTful-Drupal/restful/blob/7.x-1.x/restful.module#L567).

When you just upload a file and that's it, php://input is the raw bitstream for the file. parse_str doesn't understand it all. The code as is assumes that the Content-Type is either application/json or x-www-form-encoded. 

For reference, here is the code I'm using to make the request:

``` javascript
var config = {
  headers: {
    'Content-Type': file.type,
    'Content-Length': file.size
  }
};

$http.put(Drupal.settings.paths.api+'/files/'+scope.file.id, file, config)
```

file in this case is a [File object](https://developer.mozilla.org/en-US/docs/Web/API/File)

**Comment by amitaibu** [comment]
I understand. Alright, so if you can, please write the PR.
It would probably be also nice (could be a follow up) to make the default file resource support full crud

**Comment by e0ipso** [comment]
> The code as is assumes that the Content-Type is either application/json or x-www-form-encoded.

I think that is the key part.

We should support other content types. In the meanwhile, does a combination of `hook_restful_parse_request_alter` + `_restful_get_query_string_from_php_input` get you out of the spot? If so, would you care to contribute your code back?

Thanks!

**Comment by mpetrovic** [comment]
@mateu-aguilo-bosch While I believe that would work as well, I don't think
it should be up to the plugin writer to have to work around this code not
functioning in certain cases.

I have a patch written, it works with direct file uploads. I just need to
test it with other content types to make sure those works. I'll push it
somewhere when I get into the office.
